### PR TITLE
Persist the state of the maintenance after an upgrade

### LINKED ISSUE #16429: Remember state of maintenance:mode when running an upgrade
### Expected behavior

- `./occ maintenance:mode --on`
- Update packages
- Adjust things
- `./occ upgrade`
- Check if stuff did not explode
- `./occ maintenance:mode --off`

### Actual behavior

The upgrade mode disables maintenance mode when finished regardless on the previous state. Sure you can re-enable it, but that still leaves a few seconds where clients can hammer the server.

I've also read that there is a allow-only-admins mode. Using that for the verification would be even cooler, but I don't know how it interacts with the maintenance mode.

/cc @DeepDiver1975 @LukasReschke

### PR DESCRIPTION
**Pull request by MorrisJobke**
- if maintenance mode was enabled before an upgrade
  it will be enabled afterwards too
- fixes #16429 

cc @nickvergessen @DeepDiver1975 @danimo

### CODE REVIEW

File: core/ajax/update.php [hunk 1]
```diff
@@ -45,10 +45,10 @@
 	$disabledThirdPartyApps = [];
 
 	$updater->listen('\OC\Updater', 'maintenanceStart', function () use ($eventSource, $l) {
-		$eventSource->send('success', (string)$l->t('Turned on maintenance mode'));
+		$eventSource->send('success', (string)$l->t('Maintenance started'));
 	});
 	$updater->listen('\OC\Updater', 'maintenanceEnd', function () use ($eventSource, $l) {
-		$eventSource->send('success', (string)$l->t('Turned off maintenance mode'));
+		$eventSource->send('success', (string)$l->t('Maintenance finished'));
```
Comment (nickvergessen) [inline core/ajax/update.php]: A mode is nothing that can be finished?!

File: lib/private/updater.php [hunk 2]
```diff
@@ -163,6 +163,8 @@ public function check($updaterUrl = null) {
 	 * @return bool true if the operation succeeded, false otherwise
 	 */
 	public function upgrade() {
+		$wasMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+
 		$this->config->setSystemValue('maintenance', true);
```
Comment (nickvergessen) [inline lib/private/updater.php]: since this writes the file, wrap it with `wasMaintenanceModeEnabled` aswell?

**Comment by MorrisJobke** [comment]
The weird thing with those events is that the state change of the maintenance mode is bound to the event that the upgrade finished. Which is odd, because the maintenance mode isn't related to the upgrade state.

**Comment by MorrisJobke** [comment]
I fixed all the comments

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

File: lib/private/updater.php [hunk 3]
```diff
@@ -163,14 +163,18 @@ public function check($updaterUrl = null) {
 	 * @return bool true if the operation succeeded, false otherwise
 	 */
 	public function upgrade() {
-		$this->config->setSystemValue('maintenance', true);
+		$wasMaintenanceModeEnabled = $this->config->getSystemValue('maintenance', false);
+
+		if(!$wasMaintenanceModeEnabled) {
+			$this->config->setSystemValue('maintenance', true);
+			$this->emit('\OC\Updater', 'maintenanceEnabled');
```
Comment (nickvergessen) [inline lib/private/updater.php]: Not sure if renaming a hook (which basically is public API) is the best idea

**Comment by nickvergessen** [comment]
Other then my comment, fine by me

**Comment by PVince81** [comment]
CC @jnweiger as it might affect packaging (but probably low risk)

File: lib/private/updater.php (hunk 3 above)
Comment (DeepDiver1975) [inline lib/private/updater.php]: Generally yes - but in this case this is fine

Comment (nickvergessen) [inline lib/private/updater.php]: Well, fine by me then

**Comment by nickvergessen** [comment]
:+1:

**Comment by jnweiger** [comment]
@PVince81 Anything I should watch for in particular?

**Comment by PVince81** [comment]
@jnfrmarks it depends how the package does it.
Does it first enable maintenance and then runs the upgrade ? Does it explicitly disable maintenance mode after upgrade or does it assume that the upgrade will auto-disable it ? If the latter, then you might need to adapt the package to explicitly re-disable maintenance mode at the end.

**Comment by PVince81** [comment]
sorry, I meant @jnweiger

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**